### PR TITLE
Add CookFlavor property for projects.

### PR DESCRIPTION
**Pull request by JasperDeLaat94**
Hey there June,

Here's another merge request regarding the cook-related issue I had (see Discord for elaboration). I've made it a new target property. This now produces an Android (Game) build for me with texture compressed in a format my Quest accepts.

This one is a bit more involved, if anything needs changing I'm all ears!

Cheers!

### CODE REVIEW

File: UET/Redpoint.Uet.Configuration/Project/BuildConfigProjectBuildTarget.cs [hunk 1]
```diff
@@ -25,5 +25,11 @@ public class BuildConfigProjectBuildTarget
         [JsonPropertyName("Configurations"), JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
         [SuppressMessage("Performance", "CA1819:Properties should not return arrays", Justification = "This property is used for JSON serialization.")]
         public string[]? Configurations { get; set; } = null;
+
+        /// <summary>
+        /// The flavor to cook the project for. For Android this is texture compression, e.g. ASTC, DXT, ETC2, Multi
+        /// </summary>
+        [JsonPropertyName("CookFlavor"), JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+        public string? CookFlavor { get; set; } = null;
```
[the remark this answers is not in the document]
Comment (JasperDeLaat94) [reply UET/Redpoint.Uet.Configuration/Project/BuildConfigProjectBuildTarget.cs:33]: I think it's possible. I'll look into this when I've recovered from the flu 🥲

**Comment by JasperDeLaat94** [comment]
I've managed to implemented your suggestions, however it doesn't quite work yet. 

The pak & stage step currently seems to overwrite cook flavors when multiple are specified (e.g. "ETC2", "ASTC" results in only ASTC being present in the final .obb). Not sure why it's doing that (will have to dig a bit deeper into how staging works) but this is a good moment to commit most of the work done. Feel free to take a look, see if you spot any things that need to be corrected.

Coincidentally, I fixed AndroidClient not producing valid output, it was because CookPlatform only gets appended "Client" for Windows, Mac & Linux. Perhaps as a separate commit it should also do that for any platform that supports Client targets (I've never worked with console platforms before, so not sure there).

**Comment by hach-que** [comment]
If you want to rebase this branch on the latest `main` when you get a chance, that'll get CI/CD working for the merge.

**Comment by JasperDeLaat94** [comment]
The pak & stage step actually works as desired and produces correct output for all cook flavors:
image

It appears that if you stage multiple deployments (e.g. cook flavors, but I think also platform architectures), the Binaries folder is no longer the right place to look for all output files combined. It will only contain the output files of the last deployment:
image

It actually turns out that the AutomationTool's archive command takes care of creating dedicated output folders per deployment target: 
image
image
Note: Symbol files are not archived.

I could add `-archive -archivedirectory=<somewhere>` to the Pak & Stage step and tag those files instead, though I feel like I might be opening a big can of worms there that would break a lot of stuff. Maybe only do it for Android? 

What are your thoughts on it?

**Comment by hach-que** [comment]
Unfortunately from memory, the files that go into the archive directory aren't correct for Android builds when performing Gauntlet testing, so we still need to pull from binaries.

Is it possible for us to do the `BuildCookRun -cookflavor=` per cook flavour, rather than passing all cook flavours at once? For this to work, we'd need to pull the cook flavours into a `ForEach` node outside the `<Agent>` tag in each "Pak and Stage" step. I would do something like this:

```xml
    <ForEach Name="TargetName" Values="$(GameTargets)">
      <ForEach Name="TargetPlatform" Values="$(GameTargetPlatforms)">
        <ForEach Name="TargetConfiguration" Values="$(GameConfigurations)">
          <Property Name="EffectiveCookFlavors" Value="Unused" />
          <Property Name="EffectiveCookFlavors" Value="$(AndroidGameCookFlavors)" If="'$(StagePlatform)' == 'Android'" />
          <ForEach Name="TargetCookFlavor" Values="$(EffectiveCookFlavors)">
            <Agent Name="Pak and Stage $(TargetName) $(TargetPlatform) $(TargetConfiguration) (Windows Pak and Stage)" Type="Win64">
              <Node Name="Pak and Stage $(TargetName) $(TargetPlatform) $(TargetConfiguration)" Requires="#GameBinaries_$(TargetName)_$(TargetPlatform)_$(TargetConfiguration);#GameCookedContent_$(TargetPlatform)" Produces="#GameStaged_$(TargetName)_$(TargetPlatform)_$(TargetConfiguration)"  If="!ContainsItem('$(MacPlatforms)', '$(TargetPlatform)', ';')">
                <!-- ... -->
                <Property Name="CookFlavors" Value="" />
                <Property Name="CookFlavors" Value="$(TargetCookFlavor)" If="'$(TargetCookFlavor)' != 'Unused'" />
```

**Comment by JasperDeLaat94** [comment]
Yeah, I agree. It also allows things to run as much in parallel as possible so I went ahead and started implementing it.

Unfortunately, this does mean we now have potentially more than one staging nodes, each with distinct output. I'll have to go over the code for testing and deployment to add support for this. This will take a bit longer.

**Comment by JasperDeLaat94** [comment]
Okay, it turns out that only CustomProjectTestProvider needs changes. It currently generates a single node that requires all staged files (i.e. `"$(GameStaged);$(ClientStaged);$(ServerStaged)"`). If you build multiple Android cook flavors, then BuildGraph will throw file errors because the files are identical:

>ERROR: File 'E:\...\Binaries\Android\MyGame-arm64.apk' was produced by Pak and Stage MyGame Android Development ETC2/ClientStaged_MyGameClient_Android_Development_ETC2 and Pak and Stage MyGame Android Development ASTC/ClientStaged_MyGameClient_Android_Development_ASTC
etc, etc.

I think, in this case, it's probably best to invoke the custom test script for each individual staged build, rather than to fetch them all and then invoke the custom test script once? 

I've also considered something similar to what you implemented with Steam deployment (having a "Package" field that specifies which specific target you want the test to run for), but I think the user would rather still have their test being run for every build, rather than specific ones.

**Comment by hach-que** [comment]
Hmm, the idea of having the custom script require all the staged targets is so you can have custom scripts that start a dedicated server process and a game process and have them talk to each other (or so you can deploy game targets onto different devices at once).

Having the Binaries folder conflict is awkward though, especially since Android is most likely than most to rely on the contents of the Binaries folder.

The only thing I could think of doing is having Pak and Stage rename `<project>/Binaries/Android` to e.g. `<project>/Binaries/Android_ASTC` when multiple cook flavours are present, after the package is done, and then stage `Android_ASTC` instead of `Android`.

It's pretty janky but it should work. The only project test provider that would need updating in the future would be Gauntlet, where `BuildConfigProjectTestGauntletRequire` would have to be able to accept a cook flavour for Android, but Gauntlet for projects doesn't appear to be implemented at the moment so we don't need to worry about changing that right now.

**Comment by JasperDeLaat94** [comment]
That was easier than expected to implement (just a `<Copy>` task on the binaries after staging). As a side effect, I now get valid output folders for each cook flavor when I execute this locally, neat!

I've checked the other test and deployment providers, they are either not relevant (e.g. Steam, or test providers that only need editor binaries) or not implemented. I think with these changes the feature is ready for final review. 

Feel free to review the changes.

File: UET/Redpoint.Uet.BuildPipeline/BuildGraph/BuildGraph_Project.xml [hunk 2]
```diff
@@ -335,21 +371,33 @@
...
+                <Property Name="PackageFlag" Value="" />
+                <Property Name="PackageFlag" Value="-package" If="('$(TargetPlatform)' == 'IOS') or ('$(TargetPlatform)' == 'Android')" />
+                <Command Name="BuildCookRun" Arguments="&quot;-project=$(UProjectPath)&quot; -nop4 $(DisableCodeSign) &quot;-platform=$(TargetPlatform)&quot; $(CookFlavorArg) &quot;-clientconfig=$(TargetConfiguration)&quot; -SkipCook -cook -pak $(PackageFlag) -stage &quot;-stagingdirectory=$(StageDirectory)&quot; -unattended -stdlog -ClientCookedTargets=$(TargetName) -target=$(TargetName)" />
+                <Tag BaseDir="$(StageDirectory)/$(StagePlatform)" Files="..." With="$(StageOutputs)" />
+                <Tag BaseDir="$(ProjectRoot)/Binaries/$(TargetPlatform)" Files="..." With="$(StageOutputs)" If="'$(TargetPlatform)' == 'IOS'" />
+                <Do If="'$(TargetPlatform)' == 'Android'">
+                  <Copy From="$(ProjectRoot)\Binaries\$(TargetPlatform)\..." To="$(ProjectRoot)\Binaries\$(TargetPlatform)_$(CookFlavor)" />
```
Comment (hach-que) [inline UET/Redpoint.Uet.BuildPipeline/BuildGraph/BuildGraph_Project.xml:394]: Should these paths be e.g. `$(ProjectRoot)/Binaries/$(TargetPlatform)/...` for the Mac host nodes?

File: UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    using System.Collections.Generic;
+    using System.Text.Json.Serialization;
+    using System.Text.Json;
+
+    public class BuildConfigTargetPlatformConverter : JsonConverter<BuildConfigTargetPlatform[]>
+    {
+        public override BuildConfigTargetPlatform[] Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType != JsonTokenType.StartArray)
+                throw new JsonException("Expected StartArray token");
```
Comment (hach-que) [inline UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs:13]: Styling - Braces `{}`, including down below as well.

File: UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+            var platforms = new List<BuildConfigTargetPlatform>();
+
+            while (reader.Read() && reader.TokenType != JsonTokenType.EndArray)
+            {
+                if (reader.TokenType == JsonTokenType.String)
+                {
+                    platforms.Add(new BuildConfigTargetPlatform()
```
Comment (hach-que) [inline UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs:21]: Styling - Omit parenthesis for parameterless constructor, including down below as well.

File: UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                                platform.CookFlavors = [.. cookFlavors];
+                                break;
+                        }
+                    }
+
+                    platforms.Add(platform);
+                }
+            }
+
+            return [.. platforms];
```
Comment (hach-que) [inline UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs:60]: Styling - Guessing this is equivalent to `.ToArray()` but we're not using this `[.. var]` syntax elsewhere, so replace with `platforms.ToArray()`.

File: UET/uet/Commands/Build/DefaultBuildSpecificationGenerator.cs [hunk 6]
```diff
@@ -114,18 +115,21 @@ private static TargetConfig ComputeTargetConfig(string name, BuildConfigProjectB
...
-            var targetPlatforms = FilterIncompatiblePlatforms(target.Platforms, localExecutor);
+            var targetPlatforms = FilterIncompatiblePlatforms(target.Platforms.Select(x => x.Platform).ToArray(), localExecutor);
             var configurations = target.Configurations ?? new[] { "Development", "Shipping" };
+            var cookFlavors = target.Platforms.FirstOrDefault(x => x.Platform == "Android")?.CookFlavors ?? [];
 
             return new TargetConfig
             {
                 Targets = string.Join(";", targets),
-                TargetPlatforms = string.Join(";", targetPlatforms),
+                TargetPlatforms =  string.Join(";", targetPlatforms),
```
Comment (hach-que) [inline UET/uet/Commands/Build/DefaultBuildSpecificationGenerator.cs:130]: Styling - remove extra space.

**Review by hach-que — CHANGES_REQUESTED** [review]
A few minor style issues, plus one change in the BuildGraph XML that is probably necessary for Android builds on macOS to work.

File: UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs (hunk 5 above)
Comment (JasperDeLaat94) [reply UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs:60]: Apparently it's called a collection expression, but agree that `.ToArray()` is more readable.

File: UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs (hunk 4 above)
Comment (JasperDeLaat94) [reply UET/Redpoint.Uet.Configuration/BuildConfigTargetPlatformConverter.cs:21]: Only removed parenthesis in that instance, down below doesn't have initializers, so parentheses are required.

File: UET/Redpoint.Uet.BuildPipeline/BuildGraph/BuildGraph_Project.xml (hunk 2 above)
Comment (JasperDeLaat94) [reply UET/Redpoint.Uet.BuildPipeline/BuildGraph/BuildGraph_Project.xml:394]: Agree for consistency (copy/paste error, whoops). 

Note: When I was testing earlier (on Windows) and made a mistake, it printed an error with the path `"E:/Project/Root/Path\Binaries\Android\Subfolder/File"` which was kinda weird. I guess BuildGraph internally perhaps always tries to normalize file paths to `/` (which Windows tolerates), but yeah I can't assume that MacOS plays that nice along as well.

**Review by hach-que — APPROVED** [review]

**Comment by hach-que** [comment]
Thanks! This is now merged :)

**Pull request merged**